### DISCUSSION TITLE
Fuzzyness about rootless usage

### DISCUSSION THREAD

**Discussion by gotcha** [dagger/dagger#12177] [Help]
https://github.com/dagger/dagger/blob/6a237d77d1277951d41f307b2505eee44a110747/docs/current_docs/integrations/nerdctl.mdx?plain=1#L15

The line above seems to imply that dagger can be used with rootless containers.

However, https://github.com/dagger/dagger/blob/main/core/docs/d7yxc-operator_manual.md#can-i-run-the-dagger-engine-as-a-rootless-container states very explicitely that the Engine cannot be ran rootless.

Would it be useful to clarify that point on the `nerdctl` page ? Or did I misunderstand anything ?

**Comment by jedevc** [comment]
Good shout @gotcha - looks like we made a mistake in https://github.com/dagger/dagger/pull/8242.

Dagger doesn't support rootless, so yes, the instructions are wrong.

**Comment by gotcha** [comment]
> Dagger doesn't support rootless, so yes, the instructions are wrong.

Security conscious organizations do not like this state of affairs. Are there plans to support rootless, even far away ? Or is that a no go ?

**Comment by gotcha** [comment]
> you might be able to use rootless with a custom OCI runtime like nerdctl

The https://github.com/dagger/dagger/issues/1287#issuecomment-1673365304 also suggests to research `nerdctl` support. Should it be edited ?

**Comment by jedevc** [comment]
Rootless support is eventually possible - however, it's not on the immediate roadmap. There's quite a lot of work and plumbing we'll need to make this work. We do a lot of custom networking to enable `Service`s, and similar things.

That's not to say it's impossible (I'd definitely like to support it one day), it's that it's not currently a priority - we're a small team, we sadly don't have infinite capacity. Atm, a lot of our effort is working on improving performance, building out better observability, as well as just our regular feature work.

The research from the comment you link to was done, and then written down in the operator manual.

**Comment by shykes** [comment]
@gotcha the Dagger engine acts as a container runtime. As such it is normal for it to have system privileges required to run containers. Running dagger in a rootless container means running runc+containerd as a rootless container: not impossible but more of a science project; and not required for security.

Security-conscious teams should apply the same best practices to dagger as they do to buildkit/containerd/runc. In practice that means vm-level isolation of workloads.

**Comment by gotcha** [comment]
@shykes Thanks for the detailed answer. Not being a security specialist, your words will help me having constructive conversations with people that jump to quick conclusions based on your documentation.

I appreciated your take on this that you stated in https://www.youtube.com/watch?v=Sn1w51Vh0mM
